File: app/src/main/java/com/noteapp/Controll/NotesAdapter.java
package com.noteapp.Controll;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.noteapp.Model.Note;
import com.noteapp.R;

import java.util.ArrayList;
import java.util.List;

public class NotesAdapter extends RecyclerView.Adapter<NotesAdapter.NoteViewHolder> {

    private NoteClickListener clickListener;
    private List<Note> dataset;

    public NotesAdapter(NoteClickListener clickListener) {
        this.clickListener = clickListener;
        this.dataset = new ArrayList<>();
    }

    public void setNotes(@NonNull List<Note> notes) {
        dataset = notes;
        notifyDataSetChanged();
    }

    public Note getNote(int position) {
        return dataset.get(position);
    }

    @Override
    public NotesAdapter.NoteViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_note, parent, false);
        return new NoteViewHolder(view, clickListener);
    }

    @Override
    public void onBindViewHolder(NotesAdapter.NoteViewHolder holder, int position) {
        Note note = dataset.get(position);
        holder.titleTV.setText(note.getTitle());
        holder.textTV.setText(note.getText());
        holder.dateTV.setText(note.getDate().toString());
    }

    @Override
    public int getItemCount() {
        return dataset.size();
    }

    public interface NoteClickListener {
        void onNoteClick(int position);
    }

    static class NoteViewHolder extends RecyclerView.ViewHolder {

        public TextView titleTV, textTV, dateTV;

        public NoteViewHolder(View itemView, final NoteClickListener clickListener) {
            super(itemView);
            titleTV = (TextView) itemView.findViewById(R.id.TVNoteTitle);
            textTV = (TextView) itemView.findViewById(R.id.TVNoteText);
            dateTV = (TextView) itemView.findViewById(R.id.TVNoteDate);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (clickListener != null) {
                        clickListener.onNoteClick(getAdapterPosition());
                    }
                }
            });
        }
    }
}

File: app/src/main/java/com/noteapp/Controll/NotesAdapter.kt
package com.noteapp.Controll

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.noteapp.Model.Note
import com.noteapp.R
import java.util.*

class NotesAdapter : RecyclerView.Adapter<NoteViewHolder>() {

    private var dataset: List<Note> = ArrayList<Note>()
    private var clickListener: NoteClickListener? = null

    fun setClickListener(clickListener: NoteClickListener) {
        this.clickListener = clickListener
    }

    fun setNotes(notes: List<Note>) {
        this.dataset = notes
        notifyDataSetChanged()
    }

    fun getNote(position: Int): Note {
        return this.dataset[position]
    }

    override fun getItemCount(): Int {
        return this.dataset.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NoteViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_note, parent, false)
        return NoteViewHolder(view, clickListener, this)
    }

    override fun onBindViewHolder(holder: NoteViewHolder, position: Int) {
        val (id, title, text, date) = dataset[position]
        holder.titleTV.text = title
        holder.textTV.text = text
        holder.dateTV.text = date.toString()
    }
}